pipeline-model-api/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ast/ModelASTStage.java

```diff
@@ -112,87 +78,43 @@ public void validate(@Nonnull final ModelValidator validator) {
     public void validate(final ModelValidator validator, boolean isWithinParallel) {
         validator.validateElement(this, isWithinParallel);
 
-        if (agent != null) {
-            agent.validate(validator);
-        }
-        if (when != null) {
-            when.validate(validator);
-        }
-        if (post != null) {
-            post.validate(validator);
-        }
-        if (tools != null) {
-            tools.validate(validator);
-        }
-        if (environment != null) {
-            environment.validate(validator);
-        }
-        if (options != null) {
-            options.validate(validator);
-        }
-        if (input != null) {
-            input.validate(validator);
-        }
+        validate(validator, agent);
+        validate(validator, when);
+        validate(validator, post);
+        validate(validator, tools);
+        validate(validator, environment);
+        validate(validator, options);
+        validate(validator, input);
         if (stages != null) {
             stages.validate(validator, isWithinParallel);
         }
-        if (parallel != null) {
-            parallel.validate(validator);
-        }
-        for (ModelASTBranch branch : branches) {
-            branch.validate(validator);
-        }
-        if (matrix != null) {
-            matrix.validate(validator);
-        }
+        validate(validator, parallel);
+        validate(validator, branches);
+        validate(validator, matrix);
     }
 
     @Override
     public String toGroovy() {
         StringBuilder result = new StringBuilder();
         // TODO decide if we need to support multiline names
         result.append("stage(\'").append(name.replace("'", "\\'")).append("\') {\n");
-        if (agent != null) {
-            result.append(agent.toGroovy());
-        }
-        if (when != null) {
-            result.append(when.toGroovy());
-        }
-        if (tools != null) {
-            result.append(tools.toGroovy());
-        }
-        if (environment != null) {
-            result.append(environment.toGroovy());
-        }
-        if (options != null) {
-            result.append(options.toGroovy());
-        }
-        if (input != null) {
-            result.append(input.toGroovy());
-        }
-        if (post != null) {
-            result.append(post.toGroovy());
-        }
-        if (stages != null) {
-            result.append("stages {\n");
-            result.append(stages.toGroovy());
-            result.append("}\n");
-        }
+        result.append(toGroovy(agent));
+        result.append(toGroovy(when));
+        result.append(toGroovy(tools));
+        result.append(toGroovy(environment));
+        result.append(toGroovy(options));
+        result.append(toGroovy(input));
+        result.append(toGroovy(post));
+        result.append(toGroovy(stages));
```

Comment:
Does this still have the surrounding `stages { ... }`? Same for `parallel` and `matrix` below. Maybe I am looking at the wrong `toJSON` overload? 

Ah, ok, It looks like you moved that part down into the types themselves, i.e. `ModelASTStages.toGroovy`.